### DISCUSSION TITLE
Some questions about CTL

### DISCUSSION THREAD

**Discussion by vittorioromeo** [jart/cosmopolitan#1241] [Q&A]
Hey @jart, I've stumbled upon [`ctl`](https://github.com/jart/cosmopolitan/tree/master/ctl) and I think it's a great idea!

I hate how bloated the current Standard Library implementations are, and I have quite a bit of experience re-implementing parts of the STL to avoid compile-time overhead, see these examples from my SFML fork:

- [`sf::Optional<T>`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/Optional.hpp)
- [`sf::priv::isSame<T, U>`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/IsSame.hpp)
- [`sf::priv::UniquePtr<T>`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/UniquePtr.hpp)
- [`sf::Variant<Ts...>`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/Variant.hpp)

I might be interested in contributing to CTL :)

Now, for my questions:

1. **What C++ standard does CTL target?**

2. **What platforms/compilers does CTL target?**

3. Have you considered using `__has_builtin` to avoid template instantiations for type traits as I did in [`sf::priv::isSame<T, U>`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/IsSame.hpp)?

4. Another major issue with current Standard Library implementation is debug performance, as it's destroyed by layers of abstractions. Have you considered using attributes such as `gnu::always_inline` as I did in places such as [`sf::priv::UniquePtr<T>::operator->`](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/UniquePtr.hpp#L124C5-L128)

5. Have you considered using `extern template` + explicit template instantiations for commonly used types such as `basic_string<char>` or even `vector<basic_string<char>>`?

I'd be willing to contribute these sort of changes if you think they'd be an improvement over the status quo and if you think they fit the vision for CTL.

**Comment by jart** [comment]
Hey @vittorioromeo thanks for your patience while I was on vacation. I'm really excited to hear that you're excited about CTL! I'm getting a 404 on your code examples. If you follow the CONTRIBUTING.md guidelines then your improvements would be most welcome! We normally collaborate on Discord too, in addition to GitHub. See the invite link in the README.md file.

> What C++ standard does CTL target?

We're currently targeting `-std=gnu++20` as defined by GCC 12.3.0 which is the only compiler we use.

A better question would be, which C++ standard would you like for us to support?

For example, I'm happy to go more modern only. For example, I could upgrade superconfigure so that cosmocc uses GCC 14.1. Then we could use `-std=gnu++23` and have all the latest bells and whistles. CTL could focus only on the bleeding edge standard. Anyone who wants to be able to conformance test old standards could simply use Cosmo's LLVM libcxx which has all the necessary `#ifdef` statements for doing that. Is this something you would like?

> Have you considered using `__has_builtin` to avoid template instantiations for type traits as I did in [sf::priv::isSame<T, U>](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/IsSame.hpp)?

Yes I would very much like that. I'm simply not familiar with all the builtins and how to use them yet. Pull requests doing this are very much welcome.

> Another major issue with current Standard Library implementation is debug performance, as it's destroyed by layers of abstractions. Have you considered using attributes such as gnu::always_inline as I did in places such as [sf::priv::UniquePtr<T>::operator->](https://github.com/vittorioromeo/SFML/blob/wip_nonglobal_graphicscontext/include/SFML/System/UniquePtr.hpp#L124C5-L128)

Sounds like another great idea. Would love to see pull requests adding this! I've already added a few of my own, on an as needed basis, whenever I run into real world annoyances or suboptimal benchmark scores.

> Have you considered using extern template + explicit template instantiations for commonly used types such as basic_string<char> or even vector<basic_string<char>>?

This is also something I'd like to see happen. The biggest user of CTL so far is llamafile and llama.cpp does use wide char STL features.

> I'd be willing to contribute these sort of changes if you think they'd be an improvement over the status quo and if you think they fit the vision for CTL.

Wonderful! Everything you said sounds perfectly aligned with our vision so far. Your changes are very much welcome!
